### PR TITLE
Prevent unbounded growth of sparse tensor in add operation

### LINKED ISSUE #34964: Gradient update of a sparse matrix results in a memory leak
## 🐛 Bug

Gradient update with `parameter.add_(val, grad)` of sparse matrices results in a memory leak that is not present in dense matrices with the same computation.

Freeing cache/calling the garbage collector does not remove the problem. The result of the computation is the same for both the sparse and the dense networks. The backward pass without an update of the matrix does not cause this problem.

## To Reproduce

Take a network that computes `||Wx||_F^2` with either a dense or a sparse `W` and minimize it. The part that produces the bug is
```
with torch.no_grad():
    net.weights.add_(-lr, net.weights.grad)
```
(Updating `net.weights.data` without in-place operations results in the same behaviour)

#### Steps to reproduce the behavior:
Imports and printing functions. The script prints the currently allocated tensors and the memory stats **after** zeroing the gradient.
```
import torch
import torch.nn as nn
import gc

def print_stats(i):
    print('\nIteration %d' % i)
    print('Resources:')
    for obj in gc.get_objects():  # as in https://discuss.pytorch.org/t/how-to-debug-causes-of-gpu-memory-leaks/6741/2
        try:
            if torch.is_tensor(obj) or (hasattr(obj, 'data') and torch.is_tensor(obj.data)):
                print('\t', type(obj), obj.size())
        except:
            pass

    print('\t{0: <15}\t{1: >6} MB'.format('Allocated', torch.cuda.memory_allocated(device) // 1024 ** 2))
    print('\t{0: <15}\t{1: >6} MB'.format('Max allocated', torch.cuda.max_memory_allocated(device) // 1024 ** 2))
    print('\t{0: <15}\t{1: >6} MB'.format('Cached', torch.cuda.memory_cached(device) // 1024 ** 2))
    print('\t{0: <15}\t{1: >6} MB'.format('Max cached', torch.cuda.max_memory_cached(device) // 1024 ** 2))
```
The sparse and the dense networks are the same, but the first one uses a sparse `W` and `torch.sparse.mm(W,x)` instead of dense `W`/`torch.mm(W,x)` (`W` is `1 x n`):
```
def make_sparse_weights(input_length):
    values = torch.ones(input_length, dtype=torch.float)
    indices = torch.tensor([[0, i] for i in range(input_length)], dtype=torch.long).t()
    return torch.sparse.FloatTensor(indices, values, size=(1, input_length))


class SparseNetwork(nn.Module):
    def __init__(self, input_length):
        super().__init__()
        self.weights = nn.Parameter(make_sparse_weights(input_length),
                                    requires_grad=True)

    def forward(self, x):
        return torch.norm(torch.sparse.mm(self.weights, x)) ** 2


class DenseNetwork(nn.Module):
    def __init__(self, input_length):
        super().__init__()
        self.weights = nn.Parameter(make_sparse_weights(input_length).to_dense(), 
                                    requires_grad=True)

    def forward(self, x):
        return torch.norm(torch.mm(self.weights, x)) ** 2
```
The script does plain SGD with an option to free cache/collect garbage after zeroing the gradient:
```
def run_net(net, x, lr, device, n_iterations, 
            update_weight=True, empty_cache=False, collect_garbage=False):
    for i in range(n_iterations):
        # free the gradients and print resources

        if net.weights.grad is not None:
            net.weights.grad.detach_()
            net.weights.grad.zero_()

        if empty_cache:
            torch.cuda.empty_cache()
        if collect_garbage:
            gc.collect()

        if i % 1000 == 0:
            print_stats(i)

        # compute
        loss = net(x)
        loss.backward()

        # update
        if update_weight and net.weights.grad is not None:
            # the SGD implementation with p.data.add_() 
            # throws an error and suggests this solution:

            with torch.no_grad():
                net.weights.add_(-lr, net.weights.grad)

        if i % 1000 == 0:
            print('Loss: %f' % loss.item())
```
Dummy input:
```
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)

input_length = 1000
n_iterations = 4001
lr = 1e-6
x = torch.ones(input_length, 1, device=device)
```
1. Running the dense network behaves as expected:
```
run_net(DenseNetwork(input_length).to(device), x, lr, device, n_iterations, 
        update_weight=True, empty_cache=False, collect_garbage=False)
```
Output:
```
Iteration 0
Resources:
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

/usr/local/lib/python3.6/dist-packages/torch/distributed/distributed_c10d.py:101: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "


Iteration 1000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 18242.406250

Iteration 2000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 332.786316

Iteration 3000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 6.070851

Iteration 4000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 0.110747
```
2. The sparse network does the same computation (results in the same loss) but the allocated memory grows with iterations (the allocated tensors are the same though):
```
run_net(SparseNetwork(input_length).to(device), x, lr, device, n_iterations, 
        update_weight=True, empty_cache=False, collect_garbage=False)
```
Output:
```
Iteration 0
Resources:
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

/usr/local/lib/python3.6/dist-packages/torch/distributed/distributed_c10d.py:101: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "


Iteration 1000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    20 MB
	Max allocated  	    58 MB
	Cached         	    92 MB
	Max cached     	    92 MB
Loss: 18242.394531

Iteration 2000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    38 MB
	Max allocated  	   114 MB
	Cached         	   416 MB
	Max cached     	   416 MB
Loss: 332.785889

Iteration 3000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    57 MB
	Max allocated  	   172 MB
	Cached         	   976 MB
	Max cached     	   976 MB
Loss: 6.070839

Iteration 4000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    77 MB
	Max allocated  	   230 MB
	Cached         	  1856 MB
	Max cached     	  1856 MB
Loss: 0.110747
```
3. Freeing cache and calling the garbage collector every time the gradient is zeroed doesn't change anything:
```
run_net(SparseNetwork(input_length).to(device), x, lr, device, n_iterations, 
        update_weight=True, empty_cache=True, collect_garbage=False)
```
Output:
```
Iteration 0
Resources:
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

/usr/local/lib/python3.6/dist-packages/torch/distributed/distributed_c10d.py:101: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "


Iteration 1000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    19 MB
	Max allocated  	    60 MB
	Cached         	    38 MB
	Max cached     	    80 MB
Loss: 18242.394531

Iteration 2000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    38 MB
	Max allocated  	   115 MB
	Cached         	    50 MB
	Max cached     	   132 MB
Loss: 332.785889

Iteration 3000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    57 MB
	Max allocated  	   172 MB
	Cached         	    72 MB
	Max cached     	   188 MB
Loss: 6.070839

Iteration 4000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	    77 MB
	Max allocated  	   231 MB
	Cached         	    96 MB
	Max cached     	   276 MB
Loss: 0.110747
```
4. Finally, just doing the backward pass without the update works as expected:
```
run_net(SparseNetwork(input_length).to(device), x, lr, device, n_iterations, 
        update_weight=False, empty_cache=False, collect_garbage=False)
```
Output:
```
Iteration 0
Resources:
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

/usr/local/lib/python3.6/dist-packages/torch/distributed/distributed_c10d.py:101: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "


Iteration 1000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

Iteration 2000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

Iteration 3000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000

Iteration 4000
Resources:
	 <class 'torch.Tensor'> torch.Size([])
	 <class 'torch.Tensor'> torch.Size([1000, 1])
	 <class 'torch.nn.parameter.Parameter'> torch.Size([1, 1000])
	Allocated      	     0 MB
	Max allocated  	     0 MB
	Cached         	     2 MB
	Max cached     	     2 MB
Loss: 1000000.000000
```
## Expected behavior

The same sparse and dense computations should allocate and free memory in the same way.

## Environment

Google Colab:
```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.5.0
[conda] Could not collect
```

## Additional context

Colab was restarted for each run, so the max memory values represent the correct values.

The warning
```
/usr/local/lib/python3.6/dist-packages/torch/distributed/distributed_c10d.py:101: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
```
comes from `print_resources` and not the main script.

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen @vincentqb

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #34964

Sparse cuda add was implemented by just concatenating the indices and values for the tensor. If called repeatedly in a tight loop this will let `nnz` grow unbounded. In the worst case of  `x.add_(x)` it grows exponentially.

### CODE REVIEW

**Review by zou3519 — DISMISSED** [review]
seems like a reasonable heuristic. Does CPU sparse tensor add have this problem too?

**Comment by peterbell10** [comment]
I wasn't getting the same growth with CPU sparse tensors. Looking at the code though, `add_out_sparse_non_contiguous` also just concatenates the values, 
https://github.com/pytorch/pytorch/blob/82d58ed484eb04f894475de0551055f7e070f481/aten/src/ATen/native/sparse/SparseTensorMath.cpp#L510-L512
However, it's only triggered for sparse tensors with non-contiguous indices or value tensors. The CPU add for contiguous index & value tensors seems to do full addition.

**Comment by ezyang** [comment]
Oh, it's the uncoalesced addition situation. I'm not fundamentally opposed to some sort heuristic here, but I want to explore a few other options first. It sounds like part of the problem is that CUDA and CPU don't have equivalent coalescing behavior. Do you think it would be difficult for CUDA to be made to behave the same way as CPU (I could believe this is hard, due to CUDA's model, but it would be helpful if you could confirm.)

**Comment by ezyang** [comment]
I looked over the heuristic and I think it's pretty good.

**Comment by peterbell10** [comment]
The CPU implementation does a variation of merging two sorted lists. A similar thing could be done for coalesced inputs in CUDA using `thrust::merge_by_key` and then coalescing adjacent values. However, this needs multiple passes to do the final coalescing step. I had started implementing this but noticed this comment:

>  We deliberately choose to simply concat the indices and values tensors rather than merging them. This removes the need to synchronously fetch `nnz` at the end of the operation, at the cost of having a non-coalesced result. This trade-off is preferable for the common use-case of gradient accumulation.

One thought could be to `concat` for small nnz where synchronisation would be more costly than the kernel itself. But do a full merge for large `nnz`, avoiding the memory growth issues. Although, that's not really that much different from this PR; except that it does a `merge_by_key` instead of `sort_by_key`.

**Review by ezyang — APPROVED** [review]

**Comment by peterbell10** [comment]
@ezyang, @zou3519 is their anything left to do here?

**Comment by zou3519** [comment]
Sorry, landing this fell through the cracks. Could you rebase the PR, @peterbell10, just to get test signal again, and then I'll land this once the tests look fine?

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
Test failures look unrelated
